### PR TITLE
Bug 4318 - File truncation while transfer cause looping of timeouts

### PR DESCRIPTION
**Pull request by yurashek**
http://bugs.proftpd.org/show_bug.cgi?id=4318
Fixes bug on Solaris

### CODE REVIEW

File: src/data.c [hunk 1]
```diff
@@ -1516,6 +1516,9 @@ pr_sendfile_t pr_data_sendfile(int retr_fd, off_t *offset, off_t count) {
         session.total_raw_out += len;
 
         return -1;
+      } else { /* (len == 0) */
+	/* unexpected EOF i.e. if the file is truncated while transfer */
+	break;
```
Comment (Castaglia) [inline src/data.c]: Is this behavior specific to the Solaris `sendfile(2)` implementation?

Comment (yurashek) [reply src/data.c]: I apologize but I have made a mistake in placing the "break". I reworked it but I will need to test it a bit more.

Regarding to your question:
sendfile() on Solaris can return:
 - positive number =< "count" -> number of bytes transferred in this call
 - -1 on error (check errno)
 - 0 on EOF (no more bytes to transfer)
Will never return any negative number different of -1
If ret. value != -1 then ERRNO is undefined

file can be truncated i.e. by:
 cat /dev/null > file

**Comment by yurashek** [comment]
I needed to find a "better place" for my change outside of error section when signals are serviced.

**Comment by yurashek** [comment]
Closing pull-request. Fixed by T.J.Sanders

**Pull request closed without merging**